CommonTools/PileupAlgos/src/PuppiContainer.cc

```diff
@@ -349,3 +350,22 @@ std::vector<double> const &PuppiContainer::puppiWeights() {
   }
   return fWeights;
 }
+// ------------------------------------------------------------------------------------------
+void PuppiContainer::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<bool>("puppiDiagnostics", false);
+  desc.add<bool>("applyCHS", true);
+  desc.add<bool>("invertPuppi", false);
+  desc.add<bool>("useExp", false);
+  desc.add<double>("MinPuppiWeight", .01);
+  desc.add<double>("PtMaxNeutrals", 200.);
+  fillDescriptionsPuppiContainer(desc);
+  descriptions.addDefault(desc);
+}
+
+void PuppiContainer::fillDescriptionsPuppiContainer(edm::ParameterSetDescription &desc) {
+  edm::ParameterSetDescription algos;
+  PuppiAlgo::fillDescriptionsPuppiAlgo(algos);
+  std::vector<edm::ParameterSet> VPSetAlgos(1);
```

Comment:
Remove these two lines and let ```PuppiAlgo::fillDescriptionsPuppiAlgo``` add its parameters and VPSet's

Reply:
Hi @perrotta 
sorry I am struggling in this part, this is my best try:
```
void PuppiContainer::fillDescriptionsPuppiContainer(edm::ConfigurationDescriptions& descriptions) {

  edm::ParameterSetDescription desc;
  desc.add<bool>("puppiDiagnostics", true);
  desc.add<bool>("applyCHS", true);
  desc.add<bool>("invertPuppi", false);
  desc.add<bool>("useExp", false);
  desc.add<double>("MinPuppiWeight", .01);
  desc.add<double>("PtMaxNeutrals", 200.);

  edm::ParameterSetDescription algos;
  PuppiAlgo::fillDescriptionsPuppiAlgo(algos);
  std::vector<edm::ParameterSet> VPSetAlgos(1);
  desc.addVPSet( "algos", algos, VPSetAlgos );

  descriptions.addDefault(desc);

}
```
after I also modify the PuppiAlgos.cc as you mentioned:
```
void PuppiAlgo::fillDescriptionsPuppiAlgo(edm::ParameterSetDescription &desc) {

  edm::ParameterSetDescription puppialgos;
  puppialgos.add<int>("algoId", 5);
  puppialgos.add<bool>("useCharged", false);
  puppialgos.add<bool>("applyLowPUCorr", false);
  puppialgos.add<int>("combOpt", 5);
  puppialgos.add<double>("cone", .4);
  puppialgos.add<double>("rmsPtMin", .1);
  puppialgos.add<double>("rmsScaleFactor", 1.0);
  std::vector<edm::ParameterSet> VPSetPuppiAlgos;
  edm::ParameterSet puppiset;
  puppiset.addParameter<int>("algoId", 5);
  puppiset.addParameter<bool>("useCharged", false);
  puppiset.addParameter<bool>("applyLowPUCorr", false);
  puppiset.addParameter<int>("combOpt", 5);
  puppiset.addParameter<double>("cone", .4);
  puppiset.addParameter<double>("rmsPtMin", .1);
  puppiset.addParameter<double>("rmsScaleFactor", 1.0);
  VPSetPuppiAlgos.push_back(puppiset);
  desc.addVPSet("puppiAlgos", puppialgos, VPSetPuppiAlgos);

}
```
but I am getting a message like when I run one of the runTheMatrix tests:
```
----- Begin Fatal Exception 27-Sep-2019 10:22:20 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=PuppiProducer label='puppi'
Exception Message:
Illegal parameters found in configuration.  The parameters are named:
 'EtaMaxExtrap'
 'MedEtaSF'
 'MinNeutralPt'
 'MinNeutralPtSlope'
 'RMSEtaSF'
 'etaMax'
 'etaMin'
 'ptMin'
You could be trying to use parameter names that are not
allowed for this plugin or they could be misspelled.
```
do you have any suggestion?

Reply:
I can have a closer look later on... but are you keeping in your modified code the first lines of PuppiAlgo::fillDescriptionsPuppiAlgo, i.e. lines 238-245 of your PR? I don't see them in your code snippet here above.

Reply:
yes, I include them, in case you want to see all the files, I created another working branch: https://github.com/alefisico/cmssw/blob/puppiProducer_110_fixes/CommonTools/PileupAlgos/src/PuppiAlgo.cc

Reply:
Try

```C++
void PuppiContainer::fillDescriptionsPuppiContainer(edm::ParameterSetDescription &desc) {
  desc.add<bool>("puppiDiagnostics", false);
  desc.add<bool>("applyCHS", true);
  desc.add<bool>("invertPuppi", false);
  desc.add<bool>("useExp", false);
  desc.add<double>("MinPuppiWeight", .01);
  desc.add<double>("PtMaxNeutrals", 200.);
  PuppiAlgo::fillDescriptionsPuppiAlgo(desc);
}
```

Reply:
this is what I tried before, but then it is complaining about the "algos" parameter

Reply:
I don't see any "algos" parameter in the piece of code that I sent you...

Reply:
exactly, but the PuppiContainer has the `algo` parameter: https://github.com/alefisico/cmssw/blob/puppiProducer_110_fixes/CommonTools/PileupAlgos/src/PuppiContainer.cc#L19

Reply:
... which is called "puppiAlgos" in PuppiAlgo::fillDescriptionsPuppiAlgo

Reply:
no no, there are two parameters: one if `algos` that is a PVSet containing sets of parameters, and one of those parameters is `puppiAlgos`. Maybe it is easier if you see it here: https://github.com/cms-sw/cmssw/blob/02d4198c0b6615287fd88e9a8ff650aea994412e/CommonTools/PileupAlgos/python/Puppi_cff.py#L46-L68

Reply:
I see...
Then you should pass the whole content defined in PuppiContainer::fillDescriptionsPuppiContainer as a VPSet (called "algos")